crates/sequencer/src/commitment/controller.rs

```diff
@@ -0,0 +1,161 @@
+use std::cmp;
+
+use citrea_common::utils::merge_state_diffs;
+use citrea_primitives::compression::compress_blob;
+use citrea_primitives::MAX_TXBODY_SIZE;
+use sov_db::ledger_db::SequencerLedgerOps;
+use sov_db::schema::types::BatchNumber;
+use sov_modules_api::StateDiff;
+use tracing::{debug, warn};
+
+use super::CommitmentInfo;
+
+pub struct CommitmentController<Db>
+where
+    Db: SequencerLedgerOps,
+{
+    ledger_db: Db,
+    min_soft_confirmations: u64,
+    last_state_diff: StateDiff,
+}
+
+impl<Db> CommitmentController<Db>
+where
+    Db: SequencerLedgerOps,
+{
+    pub fn new(ledger_db: Db, min_soft_confirmations: u64) -> Self {
+        let last_state_diff = ledger_db.get_state_diff().unwrap_or_default();
```

Comment:
I am not sure if this logic is correct. What if we launch controller with sdiff from L2=1000, but .should_commit is called with L2=900? Their merging does not make sense.

Reply:
It is the responsibility of Sequencer runner to ensure that each new L2 block is sent through the channel in order. If somehow there is a mismatch in between blocks and statediff, then we have a very big problem.

The only place where we produce multiple L2 blocks, but do send only signal is when we have missed DA blocks. Which is then fine, because those are guaranteed to be empty blocks, so there will not be any state diff for sure.

It might actually be safe to store the last l2 height in CommitmentController, and assert that the next incoming request to should_commit is 1 increment of it.

WDYT?

Reply:
Good idea